elastic/fleet-server
# Support for remote ES output

Changes in this pr:
- Fleet Server handles remote_elasticsearch outputs: creates a new child bulker for each, uses this bulker to read, create, update API keys for remote ES
- if remote config changed, stop and start a new child bulker
- API keys invalidated if remote ES output is removed from agent policy
- added unit and integration tests to cover the changes

- will be added in another pr: report error state if remote ES is not accessible



## What is the problem this PR solves?

Remote ES output support in Fleet Server

## How does this PR solve the problem?

Generate API keys for agents using remote ES output host and service_token created for fleet-server-remote

## How to test this PR locally



1. Start kibana locally with the changes in [this pr](https://github.com/elastic/kibana/pull/169252)
2. Create a cloud deployment to use as remote monitoring cluster, tested with [pr deployment](https://kibana-pr-169252.kb.us-west2.gcp.elastic-cloud.com:9243/app), but can be any 8.12-SNAPSHOT
3. Start fleet-server locally with the changes in this pr (used standalone mode)
4. Create a remote ES output with a valid service token generated in the remote cluster
```
// create service token in remote kibana console
POST kbn:/api/fleet/service_tokens
{
  "remote": true
}
```
5. Create an agent policy that uses this output as monitoring output
6. Enroll an agent locally/from VM to this policy
7. Verify that the agent API key is created on the remote cluster, and the monitoring data arrives from the enrolled agent

<img width="2547" alt="image" src="https://github.com/elastic/fleet-server/assets/90178898/fae4a109-4042-4730-862a-9805a24f51a1">

<img width="2551" alt="image" src="https://github.com/elastic/fleet-server/assets/90178898/13ab4767-a6f8-43cc-ae54-424dd16419a7">


## Design Checklist




## Checklist





## Related issues



Closes https://github.com/elastic/ingest-dev/issues/1017
